<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of single-cell RNA-seq data using flashier • flashier</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of single-cell RNA-seq data using flashier">
<meta property="og:description" content="flashier">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flashier</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/willwerscheid/flashier" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of single-cell RNA-seq data using
flashier</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/willwerscheid/flashier/blob/HEAD/vignettes/flashier_single_cell.Rmd" class="external-link"><code>vignettes/flashier_single_cell.Rmd</code></a></small>
      <div class="hidden name"><code>flashier_single_cell.Rmd</code></div>

    </div>

    
    
<p>The aim of this vignette is show how flashier can be used to analyze
single-cell RNA-seq data. This vignette is modeled after the <a href="https://stephenslab.github.io/fastTopics/articles/single_cell_rnaseq_basic.html" class="external-link">fastTopics
vignette on analyzing single-cell data</a>.</p>
<p>We begin by loading the required packages. We also set the seed so
that results can be fully reproduced.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/reshape" class="external-link">reshape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/willwerscheid/flashier" class="external-link">flashier</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stephenslab.github.io/fastTopics/" class="external-link">fastTopics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="preparing-the-single-cell-data-for-flashier">Preparing the single-cell data for flashier<a class="anchor" aria-label="anchor" href="#preparing-the-single-cell-data-for-flashier"></a>
</h2>
<p>The single-cell RNA-seq data (we use the <code>pbmc_facs</code> data
from the <code>fastTopics</code> package) are unique molecular
identifier (UMI) counts stored as an <span class="math inline">\(n
\times p\)</span> sparse matrix, where <span class="math inline">\(n\)</span> is the number of cells and <span class="math inline">\(p\)</span> is the number of genes:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc_facs"</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  3774 16791</span></span></code></pre></div>
<p>(Note that other R packages, for example Seurat, use the convention
that rows are genes and columns are cells.)</p>
<p>Since the flashier model, like other linear-model-based methods
(e.g., principal components analysis), was not designed for count data,
it is recommended to first transform the data in a way that makes them
more suitable for flashier. A widely used approach is to divide the
counts by a “size factor,” add a “pseudocount,” then take the log. We
call this transformation the “shifted logarithm,” following <a href="https://doi.org/10.1038/s41592-023-01814-1" class="external-link">this paper</a>, and we
refer to the transformed counts as “shifted log counts”.</p>
<p>In practice, the shifted log counts are computed as follows: we first
divide the counts by <span class="math inline">\(\alpha s_i\)</span>,
where <span class="math inline">\(s_i\)</span> is the size factor for
cell <span class="math inline">\(i\)</span>, and <span class="math inline">\(\alpha\)</span> is the pseudocount. Done this way,
the shifted log counts maintain sparsity in the data; that is, if the
original count is zero, then the shifted log count is also zero.</p>
<p>More formally, the transformed counts <span class="math inline">\(y_{ij}\)</span> obtained from the original counts
<span class="math inline">\(x_{ij}\)</span> are <span class="math display">\[
y_{ij} = \log\bigg(1 + \frac{x_{ij}}{\alpha s_i}\bigg).
\]</span> In our analysis below, we use library-size normalization; that
is, we set <span class="math display">\[
s_i =
\frac{\text{total count for cell} \; i}
     {\text{average across all cells}} =
\frac{\sum_{j=1}^p x_{ij}}
     {\frac{1}{n}\sum_{i=1}^n \sum_{j=1}^p x_{ij}},
\]</span> and we set the pseudocount <span class="math inline">\(\alpha\)</span> to 1.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">size_factors</span> <span class="op">&lt;-</span> <span class="va">size_factors</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">size_factors</span><span class="op">)</span></span>
<span><span class="va">shifted_log_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">/</span> <span class="op">(</span><span class="va">a</span> <span class="op">*</span> <span class="va">size_factors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that, since the counts are sparse, we could have performed this
transformation more efficiently using, for example, the “mapSparse”
function from <a href="https://github.com/david-cortes/MatrixExtra" class="external-link">the
MatrixExtra package</a>.</p>
</div>
<div class="section level2">
<h2 id="variance-regularization">Variance regularization<a class="anchor" aria-label="anchor" href="#variance-regularization"></a>
</h2>
<p>Before we run flashier, one issue we need to confront is that
flashier may automatically estimate the variances to be too small, which
can especially be an issue for genes with low expression. We can avoid
this issue by setting a sensible lower bound on the variance
estimates.</p>
<!-- When running flashier, we need to place some regularization on the -->
<!-- variance estimates. Without regularization, it is possible for some -->
<!-- variance estimates to be very small, which can lead to overfitting. -->
<p>We use the following rule of thumb: estimate the standard deviation
of the transformed data for a Poisson random variable with rate <span class="math inline">\(\mu = 1 / n\)</span>, where <span class="math inline">\(n\)</span> is the number of samples. This standard
deviation corresponds to a gene for which we would expect to observe a
single count across all <span class="math inline">\(n\)</span> cells, so
it can serve as a reasonable lower bound to prevent variance estimates
from getting too small.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">x</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1e7</span>, <span class="fl">1</span><span class="op">/</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-a-flashier-model">Fit a flashier model<a class="anchor" aria-label="anchor" href="#fit-a-flashier-model"></a>
</h2>
<p>Now we can call flash to fit a flashier model to the transformed
data:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flash.html">flash</a></span><span class="op">(</span><span class="va">shifted_log_counts</span>, </span>
<span>             ebnm_fn <span class="op">=</span> <span class="va">ebnm_point_exponential</span>,</span>
<span>             var_type <span class="op">=</span> <span class="fl">2</span>, </span>
<span>             greedy_Kmax <span class="op">=</span> <span class="fl">8</span>, </span>
<span>             S <span class="op">=</span> <span class="va">s1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Adding factor 1 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 2 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 3 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 4 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 5 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 6 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 7 to flash object...</span></span>
<span><span class="co">#&gt; Adding factor 8 to flash object...</span></span>
<span><span class="co">#&gt; Wrapping up...</span></span>
<span><span class="co">#&gt; Done.</span></span>
<span><span class="co">#&gt; Nullchecking 8 factors...</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>A few notes about this flash call:</p>
<ul>
<li><p><code>ebnm_fn = ebnm_point_exponential</code> forces both the
<span class="math inline">\({\bf L}\)</span> and <span class="math inline">\({\bf F}\)</span> matrices in flashier to be
non-negative, so this call will generate a non-negative matrix
factorization (NMF) of the (transformed) counts matrix. Other types of
matrix factorizations can be produced with different choices of the
<code>ebnm_fn</code> argument; this is explained in detail in the
“Introduction to flashier” vignette.</p></li>
<li><p><code>var_type = 2</code> means that we estimate column-wise
(here, gene-wise) variances; that is, we estimate a different variance
for each column (gene).</p></li>
<li><p><code>greedy_Kmax = 8</code> forces flashier to fit no more than
8 factors. <strong>In practice we recommend using a larger
value,</strong> but to keep the example simple and short we set it to 8
here.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="visualizing-the-memberships">Visualizing the memberships<a class="anchor" aria-label="anchor" href="#visualizing-the-memberships"></a>
</h2>
<!-- Each cell in the single-cell data set is represented as a linear
combination of the $K$ factors. -->
<p>The estimated “membership levels”, or “memberships”, are stored in
the <span class="math inline">\({\bf L}\)</span> matrix. For
example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GATATATGTCAGTG-1-b_cells"</span>,</span>
<span>              <span class="st">"GACAGTACCTGTGA-1-memory_t"</span>,</span>
<span>              <span class="st">"TGAAGCACACAGCT-1-b_cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">L_pm</span><span class="op">[</span><span class="va">cell_ids</span>, <span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            [,1] [,2] [,3] [,4]  [,5]  [,6] [,7] [,8]</span></span>
<span><span class="co">#&gt; GATATATGTCAGTG-1-b_cells  0.562    0    0    0 0.001 0.487    0    0</span></span>
<span><span class="co">#&gt; GACAGTACCTGTGA-1-memory_t 0.636    0    0    0 0.247 0.000    0    0</span></span>
<span><span class="co">#&gt; TGAAGCACACAGCT-1-b_cells  0.617    0    0    0 0.000 0.211    0    0</span></span></code></pre></div>
<p>The first and third cells mainly have membership in factors 1 and 6,
whereas the second cell mostly has membership in factors 1 and 5. Most
of the memberships are zero.</p>
<p>We can use a “structure plot” to visualize the relationship between
the memberships and the cell labels, and possibly get some clues about
the biological meaning of the factors. The cell labels are the “sorted”
cell subpopulations:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span><span class="op">)</span></span>
<span><span class="co">#&gt;  B cell   CD14+   CD34+ NK cell  T cell </span></span>
<span><span class="co">#&gt;     767     163     687     673    1484</span></span></code></pre></div>
<p>A structure plot is a stacked bar plot in which each factor is
represented as a bar of a different color, and the bar heights are given
by the membership levels. For a more effective visualization, we
recommend extracting a rescaled version of the memberships using the
<code>ldf</code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ldf.html">ldf</a></span><span class="op">(</span><span class="va">fit</span>, type <span class="op">=</span> <span class="st">"i"</span><span class="op">)</span></span></code></pre></div>
<p>(Note that we can multiply the <span class="math inline">\(k\)</span>th column of <span class="math inline">\({\bf F}\)</span> by any constant scalar <span class="math inline">\(c &gt; 0\)</span> while also multiplying the <span class="math inline">\(k\)</span>th column of <span class="math inline">\({\bf L}\)</span> by <span class="math inline">\(1/c\)</span> without changing the low-rank
representation, <span class="math inline">\({\bf L}{\bf
F}^{\top}\)</span>.)</p>
<p>Now we can visualize the rescaled memberships in a structure plot. To
do this, we create a simple wrapper function that converts the flashier
results into an object that is usable by the “structure_plot” function
in fastTopics:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flashier_structure_plot</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">ldf</span>, <span class="va">topics</span>, <span class="va">grouping</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">ldf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"multinom_topic_model_fit"</span>, <span class="st">"list"</span><span class="op">)</span></span>
<span>  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stephenslab.github.io/fastTopics/reference/structure_plot.html" class="external-link">structure_plot</a></span><span class="op">(</span><span class="va">ldf</span>, topics <span class="op">=</span> <span class="va">topics</span>, grouping <span class="op">=</span> <span class="va">grouping</span>, gap <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"membership"</span>, color <span class="op">=</span> <span class="st">"factor"</span>, fill <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">flashier_structure_plot</span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">8</span>, <span class="fl">1</span><span class="op">)</span>, <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/structure-plot-1-1.png" width="840" style="display: block; margin: auto;"></p>
<p>The first factor is present in all cell types, and is likely
capturing a “baseline” level of expression throughout. To focus on
differences between cell types, let’s remove this first factor from our
structure plot:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">flashier_structure_plot</span><span class="op">(</span><span class="va">res</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">8</span>, <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/structure-plot-2-1.png" width="840" style="display: block; margin: auto;"></p>
<p>It is quite clear from this structure plot that factors 2 through 6
are capturing, respectively, natural killer (NK) cells, CD14+ cells,
CD34+ cells, T cells and B cells. There is also more subtle structure
captured by the memberships, such as the subset of T cells with mixed
memberships (factors 2 and 5, colored blue and orange in the plot),
which suggest a function that is intermediate between that of the “pure”
T and the “pure” NK cells. It is also interesting to note the subset of
CD34+ cells that are primarily loaded on factors 3 or 6 (yellow and
green); some of these cells may be mislabeled.</p>
<p>Note that factors 7 and 8 are present in very few cells and probably
offer little insight:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">L</span> <span class="op">&gt;</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3774  825  239  630 1420  826   20   18</span></span></code></pre></div>
<!-- Others may have a hybrid function. -->
<p>We can also visualize the memberships with a heatmap:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flashier_heatmap</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">L</span>, <span class="va">grouping</span>, <span class="va">gap</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span></span>
<span>  <span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"k"</span>,<span class="fl">1</span><span class="op">:</span><span class="va">k</span><span class="op">)</span></span>
<span>  <span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">grouping</span><span class="op">)</span></span>
<span>  <span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">L</span><span class="op">[</span><span class="va">rows</span>,<span class="op">]</span><span class="op">)</span>,variable_name <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span></span>
<span>  <span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pdat</span>,grouping <span class="op">=</span> <span class="va">grouping</span><span class="op">[</span><span class="va">rows</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pdat</span>,pos <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span> <span class="op">+</span> <span class="va">gap</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">pdat</span><span class="op">$</span><span class="va">grouping</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">factor</span>,y <span class="op">=</span> <span class="va">pos</span>,fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>,high <span class="op">=</span> <span class="st">"darkblue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="va">pdat</span><span class="op">$</span><span class="va">pos</span>,<span class="va">pdat</span><span class="op">$</span><span class="va">grouping</span>,<span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">""</span>,fill <span class="op">=</span> <span class="st">"membership"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">flashier_heatmap</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">L</span>,<span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span>,gap <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/heatmap-1.png" width="390" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="interpreting-the-factors-by-looking-at-changes-in-expression">Interpreting the factors by looking at changes in expression<a class="anchor" aria-label="anchor" href="#interpreting-the-factors-by-looking-at-changes-in-expression"></a>
</h2>
<p>Above, we used the provided cell labels to interpret five of the
factors as capturing distinct cell types (B cells, T cells, etc).
However, one might not always have cell labels, or the existing cell
labels may not be informative for the factors. We can also look to the
<span class="math inline">\({\bf F}\)</span> matrix for interpreting the
factors.</p>
<p>The <span class="math inline">\({\bf F}\)</span> matrix contains the
estimated changes in expression, specifically changes in expression on
the log scale (multiplicative changes). (The exception to this is the
first factor, which in this example is capturing a “baseline” level of
expression.) Intuitively, genes that were estimated to have the largest
increases in expression might give the most helpful clues about the
underlying biology. For example, many of the genes with the largest
expression increases in factor 6 are also genes characteristic of B
cells:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="cn">F</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">res</span>,<span class="cn">F</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">D</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">symbol</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="cn">F</span><span class="op">[</span>,<span class="fl">6</span><span class="op">]</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,n <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="co">#&gt;      CD74   HLA-DRA  HLA-DRB1     CD79A  HLA-DPB1  HLA-DPA1     CD79B  HLA-DRB5 </span></span>
<span><span class="co">#&gt; 3.4724286 2.9144905 2.3949018 2.2259586 2.1084118 1.9673726 1.8776798 1.6999751 </span></span>
<span><span class="co">#&gt;      CD37       LTB  HLA-DQA1  HLA-DQA2     MS4A1      CD52  HLA-DQB1     TCL1A </span></span>
<span><span class="co">#&gt; 1.6019631 1.4980763 1.3626036 1.1202287 1.1135809 1.0048497 0.9848164 0.9128516</span></span></code></pre></div>
<p>Let’s now apply this simple rule-of-thumb and examine the top 4 genes
(by expression increase) for factors 2 through 6:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_top_genes</span> <span class="op">&lt;-</span> <span class="kw">function</span> <span class="op">(</span><span class="va">F</span>, <span class="va">n</span> <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="cn">F</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="cn">F</span>,<span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">x</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,n <span class="op">=</span> <span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">rows</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">n</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rows</span><span class="op">)</span></span>
<span>  <span class="cn">F</span>    <span class="op">&lt;-</span> <span class="cn">F</span><span class="op">[</span><span class="va">rows</span>,<span class="op">]</span></span>
<span>  <span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape/man/melt-24.html" class="external-link">melt</a></span><span class="op">(</span><span class="cn">F</span>,variable_name <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span></span>
<span>  <span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pdat</span>,pos <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">factor</span>,y <span class="op">=</span> <span class="va">pos</span>,fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.8</span>,color <span class="op">=</span> <span class="st">"white"</span>,linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>,high <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span>,labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"k"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">plot_top_genes</span><span class="op">(</span><span class="cn">F</span><span class="op">[</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="flashier_single_cell_files/figure-html/top-genes-all-1.png" width="300" style="display: block; margin: auto;"></p>
<p>Indeed, this approach reveals genes characteristic of the other cell
types (e.g., GNLY for NK cells, S100A9 for CD14+ cells), although not
always; for example, some genes show large expression increases more
than one factor.</p>
<p>We caution that while this simple rule-of-thumb—annotating the
factors by the genes with the largest increases in expression—worked
well in this example, it may not always be helpful.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.4.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] cowplot_1.1.3      ggplot2_3.5.0      fastTopics_0.6-165 flashier_1.0.23   </span></span>
<span><span class="co">#&gt; [5] magrittr_2.0.3     ebnm_1.1-25        reshape_0.8.9      Matrix_1.6-5      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.4       softImpute_1.4-1   xfun_0.42          bslib_0.6.1       </span></span>
<span><span class="co">#&gt;  [5] htmlwidgets_1.6.4  ggrepel_0.9.5      lattice_0.22-5     quadprog_1.5-8    </span></span>
<span><span class="co">#&gt;  [9] vctrs_0.6.5        tools_4.3.3        generics_0.1.3     parallel_4.3.3    </span></span>
<span><span class="co">#&gt; [13] tibble_3.2.1       fansi_1.0.6        highr_0.10         pkgconfig_2.0.3   </span></span>
<span><span class="co">#&gt; [17] data.table_1.15.2  SQUAREM_2021.1     desc_1.4.3         RcppParallel_5.1.7</span></span>
<span><span class="co">#&gt; [21] lifecycle_1.0.4    truncnorm_1.0-9    farver_2.1.1       compiler_4.3.3    </span></span>
<span><span class="co">#&gt; [25] stringr_1.5.1      progress_1.2.3     textshaping_0.3.7  munsell_0.5.0     </span></span>
<span><span class="co">#&gt; [29] htmltools_0.5.7    sass_0.4.8         lazyeval_0.2.2     yaml_2.3.8        </span></span>
<span><span class="co">#&gt; [33] plotly_4.10.4      crayon_1.5.2       pillar_1.9.0       pkgdown_2.0.7     </span></span>
<span><span class="co">#&gt; [37] jquerylib_0.1.4    tidyr_1.3.1        uwot_0.1.16        cachem_1.0.8      </span></span>
<span><span class="co">#&gt; [41] trust_0.1-8        gtools_3.9.5       tidyselect_1.2.0   digest_0.6.34     </span></span>
<span><span class="co">#&gt; [45] Rtsne_0.17         stringi_1.8.3      dplyr_1.1.4        purrr_1.0.2       </span></span>
<span><span class="co">#&gt; [49] ashr_2.2-64        labeling_0.4.3     splines_4.3.3      fastmap_1.1.1     </span></span>
<span><span class="co">#&gt; [53] grid_4.3.3         colorspace_2.1-0   cli_3.6.2          invgamma_1.1      </span></span>
<span><span class="co">#&gt; [57] utf8_1.2.4         withr_3.0.0        prettyunits_1.2.0  scales_1.3.0      </span></span>
<span><span class="co">#&gt; [61] horseshoe_0.2.0    httr_1.4.7         rmarkdown_2.26     deconvolveR_1.2-1 </span></span>
<span><span class="co">#&gt; [65] hms_1.1.3          ragg_1.2.7         pbapply_1.7-2      memoise_2.0.1     </span></span>
<span><span class="co">#&gt; [69] evaluate_0.23      knitr_1.45         viridisLite_0.4.2  irlba_2.3.5.1     </span></span>
<span><span class="co">#&gt; [73] rlang_1.1.3        Rcpp_1.0.12        mixsqp_0.3-54      glue_1.7.0        </span></span>
<span><span class="co">#&gt; [77] jsonlite_1.8.8     R6_2.5.1           plyr_1.8.9         systemfonts_1.0.6 </span></span>
<span><span class="co">#&gt; [81] fs_1.6.3</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jason Willwerscheid, Peter Carbonetto, Wei Wang, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
